config/gradle/jre.gradle

```diff
@@ -1,46 +1,130 @@
+def jreVersion = '8u212'
+def jreUrlBase = "https://download.bell-sw.com/java/$jreVersion/bellsoft-jre$jreVersion"
+def jreUrlFilenames = [
+        Linux64   : 'linux-amd64.tar.gz',
+        Linux32   : 'linux-i586.tar.gz',
+        Windows64 : 'windows-amd64.zip',
+        Windows32 : 'windows-i586.zip',
+        Mac       : 'macos-amd64.zip'
+]
+
+task downloadJreAll {
+    group 'Download'
+    description 'Downloads JRE for all platforms'
 }
 
-task fetchJreAll {
-    group "Download"
-    description "Downloads JRE for all platforms"
+task allDistZip {
+    group 'Distribution'
+    description 'Bundles the project with a JRE for each platform'
+    dependsOn distZip
 }
 
-Jre.each { arch ->
-    def downloadTask = task("fetchJre${arch}", type: Download) {
-        group "Download"
-        description "Download JRE for ${arch}"
+jreUrlFilenames.each { os, file ->
+    def packedJre = new File("$projectDir/jre/$file")
+    def unpackedJre = new File("$buildDir/jre/$os")
 
-        src arch.downloadUrl
-        dest file("${project.projectDir}/jres/${arch.targetArchive}")
-        overwrite false
-        onlyIfModified true
+    def downloadTask = task("downloadJre$os") {
+        group 'Download'
+        description "Downloads JRE for $os"
+
+        doFirst {
+            download {
+                src "$jreUrlBase-$file"
+                dest packedJre
+                overwrite false
+            }
+        }
+
+        doLast {
+            // Unpack the JRE
+            if (!unpackedJre.exists()) {
+                unpackedJre.mkdirs()
+                copy {
+                    from(file.endsWith("zip")
+                            ? zipTree(packedJre)
+                            : tarTree(packedJre)) {
+                        eachFile { fcd ->
+                            fcd.relativePath = new RelativePath(
+                                    true, fcd.relativePath.segments.drop(1))
+                        }
+                        includeEmptyDirs = false
+                    }
+                    into unpackedJre
+                }
+            }
+        }
     }
 
-    fetchJreAll.dependsOn downloadTask
+    def distName = os.toLowerCase()
+    distributions {
+        "$distName" {
+            contents {
+                with distributions.main.contents
+                into('jre') {
+                    from unpackedJre
+                }
+            }
+        }
+    }
+
+    def zipTask = tasks.named("${distName}DistZip").get()
+    def tarTask = tasks.named("${distName}DistTar").get()
+
+    downloadJreAll.dependsOn downloadTask
+    zipTask.dependsOn downloadTask
+    zipTask.description "Bundles the project with a JRE for $os"
+    tarTask.dependsOn downloadTask
+    tarTask.description "Bundles the project with a JRE for $os"
+    allDistZip.dependsOn zipTask
+}
+
+/**
+ * Modify start scripts to use bundled JRE when
+ * available, otherwise fall back to searching.
+ */
+startScripts.doLast {
```

Comment:
This looks somewhat hacky, although I don't know any better way right now.. Can't we set the `JAVA_HOME` or `JAVACMD` just for the process in the start scripts?

@jdrueckert suggested to apply a patch set instead of search-and-replace...

Reply:
The template files can be found here (for Gradle v5.4.1): https://github.com/gradle/gradle/tree/v5.4.1/subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins